File: test/cancel_denial_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mugam_flutter/core/agreements/agreement_cancel.dart';

// N45, вторая половина — почему отказано.
//
// До 07.08 экран объяснял ЛЮБОЙ `permission-denied` гонкой сторон: «вас
// опередили». Пока правила отмены не были выложены, та же ветка ловила
// отсутствие правила и рассказывала про него неправду — молча, потому что
// гонка в Crashlytics не сообщается по замыслу.
//
// Сама ошибка причин не различает и не будет: Firestore не называет ни
// правила, ни условия. Различимо состояние — и правило ниже задаёт ему
// один вопрос: объясняет ли оно отказ.
//
// Тесты идут ПО ХОДАМ: у каждого своё «объясняющее» состояние, и спутать
// их нельзя — отзыв объясняется не тем же, чем подтверждение.

const me = 'me-uid';
const other = 'other-uid';

CancelDenial verdict(
  String deed, {
  String status = 'agreed',
  String? requestedBy,
}) =>
    explainCancelDenial(
      deed: deed,
      status: status,
      cancelRequestedBy: requestedBy,
      currentUid: me,
    );

void main() {
  group('отменённый договор объясняет отказ любому ходу', () {
    // Самая частая гонка: вторая сторона подтвердила отмену, пока человек
    // читал экран. Правила запрещают при `cancelled` вообще всё.
    for (final deed in [
      kCancelRequested,
      kCancelConfirmed,
      kCancelWithdrawn,
      kCancelDeclined,
    ]) {
      test('ход $deed', () {
        expect(verdict(deed, status: 'cancelled'), CancelDenial.race);
      });
    }
  });

  group('просил отмену', () {
    test('запрос уже стоит — гонка', () {
      expect(verdict(kCancelRequested, requestedBy: other), CancelDenial.race);
    });

    test('запроса нет, а отказали — причина НЕ в гонке', () {
      // Ровно тот случай, что жил в проде: правила не выложены, поле
      // пустое, отказ настоящий, а экран говорил «вас опередили».
      expect(verdict(kCancelRequested), CancelDenial.unknown);
    });
  });

  group('подтверждал чужой запрос', () {
    test('запрос отозван — гонка', () {
      expect(verdict(kCancelConfirmed), CancelDenial.race);
    });

    test('запрос оказался моим — тоже гонка', () {
      // Второй отозвал, я успел запросить сам: подтверждать собственный
      // запрос правила не дают, и это законный исход.
      expect(verdict(kCancelConfirmed, requestedBy: me), CancelDenial.race);
    });

    test('чужой запрос на месте, а отказали — НЕ гонка', () {
      expect(
        verdict(kCancelConfirmed, requestedBy: other),
        CancelDenial.unknown,
      );
    });
  });

  group('отклонял чужой запрос', () {
    test('запрос отозван — гонка', () {
      expect(verdict(kCancelDeclined), CancelDenial.race);
    });

    test('чужой запрос на месте, а отказали — НЕ гонка', () {
      expect(
        verdict(kCancelDeclined, requestedBy: other),
        CancelDenial.unknown,
      );
    });
  });

  group('отзывал свой запрос', () {
    test('запрос больше не мой — гонка', () {
      expect(verdict(kCancelWithdrawn, requestedBy: other), CancelDenial.race);
      expect(verdict(kCancelWithdrawn), CancelDenial.race);
    });

    test('мой запрос на месте, а отказали — НЕ гонка', () {
      expect(verdict(kCancelWithdrawn, requestedBy: me), CancelDenial.unknown);
    });
  });

  group('ходы не путаются между собой', () {
    test('одно состояние даёт РАЗНЫЕ ответы разным ходам', () {
      // Стоящий чужой запрос: для «просил» это гонка (опередили), для
      // «подтверждал» — нет, подтверждать было что. Правило, отвечающее
      // одинаково всем, прошло бы половину тестов выше.
      expect(verdict(kCancelRequested, requestedBy: other), CancelDenial.race);
      expect(
        verdict(kCancelConfirmed, requestedBy: other),
        CancelDenial.unknown,
      );
    });
  });

  group('граница названа вслух и не «улучшается»', () {
    test('разрешившаяся гонка попадает в неизвестный отказ', () {
      // Второй отозвал своё, пока мы получали отказ: состояние уже не
      // объясняет ничего. Ошибка в БЕЗОПАСНУЮ сторону — осторожные слова
      // и запись в Crashlytics вместо уверенной неправды. Обратный
      // размен возвращает исходный дефект.
      expect(verdict(kCancelConfirmed, requestedBy: other),
          CancelDenial.unknown);
    });

    test('неизвестный ход не выдаётся за гонку', () {
      expect(verdict('что-то новое'), CancelDenial.unknown);
    });
  });
}

File: functions/test/cancel-denial.test.ts
import fs from "fs";
import path from "path";
import assert from "node:assert";
import {
  initializeTestEnvironment,
  assertFails,
  RulesTestEnvironment,
} from "@firebase/rules-unit-testing";
import { doc, getDoc, setDoc, updateDoc } from "firebase/firestore";
import { PROJECT_ID, FIRESTORE_EMULATOR_PORT } from "./helpers";

// N45, вторая половина — ПАРНАЯ проверка различителя «гонка или нет».
//
// Правило разбора живёт на клиенте (`core/agreements/agreement_cancel.dart`,
// `explainCancelDenial`) и проверено юнит-тестами по состояниям. Здесь
// проверяется то, чего юнит-тест увидеть не может: что оба состояния
// РЕАЛЬНО ВОЗНИКАЮТ на настоящих правилах, и возникают по разным
// причинам, дающим один и тот же `permission-denied`.
//
// Пара, а не один случай, — ровно по разбору посева N49: набор, где
// сеется только «сломанное» состояние, не отличает работающее правило от
// неработающего.
//
//   А. Правила БЕЗ функции снятия запроса (так прод и жил до 06.08):
//      клиентская форма записи отклоняется, а состояние документа
//      остаётся НЕТРОНУТЫМ — запрос по-прежнему мой. Для `explainCancelDenial`
//      это «неизвестный отказ»: осторожные слова плюс Crashlytics.
//
//   Б. Настоящие правила, но поле уже занято второй стороной: та же
//      форма записи отклоняется, а состояние объясняет отказ — гонка,
//      и сообщение про «вас опередили» верно.

const OWNER = "owner";
const CONTACT = "contactUid";
const EVENT = "ev-denial";

/** Правила, из которых вынута дорога отзыва: `clearsCancelFields` всегда
 *  ложна. Ровно то, чем прод был до выкладки 06.08 — функции не
 *  существовало вовсе, и клиент получал отказ на законный ход. */
function rulesWithoutWithdraw(source: string): string {
  const marker = "function clearsCancelFields() {";
  const start = source.indexOf(marker);
  assert.ok(start > -1, "clearsCancelFields не найдена — правила изменились");
  const end = source.indexOf("\n      }", start) + "\n      }".length;
  return (
    source.slice(0, start) +
    "function clearsCancelFields() { return false; }" +
    source.slice(end)
  );
}

async function seedAgreement(env: RulesTestEnvironment, requestedBy: string) {
  await env.withSecurityRulesDisabled(async (context) => {
    await setDoc(doc(context.firestore(), `personalEvents/${EVENT}`), {
      ownerUid: OWNER,
      musicians: [OWNER, CONTACT],
      partnerUid: CONTACT,
      date: "2026-09-01T19:00:00.000",
      type: "Toy",
      location: "",
      notes: "",
      isAgree: true,
      status: "agreed",
      cancelRequestedBy: requestedBy,
      cancelRequestedAt: new Date(),
      cancelConfirmedBy: null,
      cancelledAt: null,
    });
  });
}

/** Форма записи ровно та, что шлёт клиент (`withdrawAgreementCancel`). */
const withdraw = (uid: string) => ({
  cancelRequestedBy: null,
  cancelRequestedAt: null,
  lastActionBy: uid,
  lastActionType: "cancelWithdrawn",
});

describe("N45: отказ по правам — две причины, одна ошибка", () => {
  const rulesPath = path.resolve(__dirname, "../../firestore.rules");
  const realRules = fs.readFileSync(rulesPath, "utf8");

  let real: RulesTestEnvironment;
  let stripped: RulesTestEnvironment;

  beforeAll(async () => {
    real = await initializeTestEnvironment({
      projectId: PROJECT_ID,
      firestore: {
        host: "localhost",
        port: FIRESTORE_EMULATOR_PORT,
        rules: realRules,
      },
    });
    stripped = await initializeTestEnvironment({
      projectId: `${PROJECT_ID}-stripped`,
      firestore: {
        host: "localhost",
        port: FIRESTORE_EMULATOR_PORT,
        rules: rulesWithoutWithdraw(realRules),
      },
    });
  });

  afterAll(async () => {
    await real?.cleanup();
    await stripped?.cleanup();
  });

  it("А. правила без функции: отказ есть, а состояние его НЕ объясняет", async () => {
    await seedAgreement(stripped, OWNER);
    const db = stripped.authenticatedContext(OWNER).firestore();

    await assertFails(
      updateDoc(doc(db, `personalEvents/${EVENT}`), withdraw(OWNER)),
    );

    // Состояние, которое клиент прочитает С СЕРВЕРА после отказа.
    // Запрос по-прежнему мой, договор в силе — объяснить отказ этим
    // нечем. `explainCancelDenial` обязан ответить `unknown`.
    await stripped.withSecurityRulesDisabled(async (context) => {
      const snap = await getDoc(
        doc(context.firestore(), `personalEvents/${EVENT}`),
      );
      assert.equal(snap.data()?.cancelRequestedBy, OWNER);
      assert.equal(snap.data()?.status, "agreed");
    });
  });

  it("Б. настоящие правила, чужой запрос: отказ ОБЪЯСНЯЕТСЯ состоянием", async () => {
    // Запрос подала вторая сторона — отозвать его владелец не может, и
    // это законный исход, а не поломка.
    await seedAgreement(real, CONTACT);
    const db = real.authenticatedContext(OWNER).firestore();

    await assertFails(
      updateDoc(doc(db, `personalEvents/${EVENT}`), withdraw(OWNER)),
    );

    await real.withSecurityRulesDisabled(async (context) => {
      const snap = await getDoc(
        doc(context.firestore(), `personalEvents/${EVENT}`),
      );
      // Запрос не мой — ровно то состояние, по которому клиентское
      // правило скажет «гонка» и покажет предметное сообщение.
      assert.equal(snap.data()?.cancelRequestedBy, CONTACT);
      assert.equal(snap.data()?.status, "agreed");
    });
  });

  it("пара имеет смысл: отказ в обоих случаях ОДИН И ТОТ ЖЕ", async () => {
    // Если бы причины различались самой ошибкой, весь разбор был бы не
    // нужен. Здесь оба хода отклонены одинаково — различает их только
    // состояние документа, прочитанное следом.
    await seedAgreement(stripped, OWNER);
    await seedAgreement(real, CONTACT);

    const a = stripped.authenticatedContext(OWNER).firestore();
    const b = real.authenticatedContext(OWNER).firestore();

    let codeA = "";
    let codeB = "";
    try {
      await updateDoc(doc(a, `personalEvents/${EVENT}`), withdraw(OWNER));
    } catch (e) {
      codeA = (e as { code?: string }).code ?? "";
    }
    try {
      await updateDoc(doc(b, `personalEvents/${EVENT}`), withdraw(OWNER));
    } catch (e) {
      codeB = (e as { code?: string }).code ?? "";
    }

    assert.ok(codeA.includes("permission-denied"), `А: ${codeA}`);
    assert.equal(codeA, codeB);
  });
});
